### PR TITLE
Fix quoting/escaping bugs in analyzedb

### PR DESCRIPTION
**Pull request by hlinnaka**
This is a more comprehensive fix to the quoting/escaping issues that lead to pull request #44. With this patch, analyzedb will no longer choke on tables that contain funny characters in their names.

The behaviour is slightly different than that proposed in #44. If you pass -t Foo.Bar, without any quoting, the server will resolve that using the usual lowercase-smashing rule, and will search for "foo.bar", in lowercase. To avoid that, table names should be quoted. This is the same behaviour as e.g. the Postgres' reindexdb and vacuumdb tools have, so I think that's reasonable. The -s option behaves differently though: case is preserved.

When table names are printed to the user, I tried to follow the approach that "simple" table names that don't contain any funny characters are printed as is, without quoting, but others are quoted. For example:
> foo.table
> "my schema".table
> "CamelSchema"."my table"

This still isn't a complete fix: the file formats use comma as a separator, so you'll still get trouble if your table or schema name contain a comma. We really ought to fix that too, but it's orthogonal enough (i.e. it will not require rewriting the changes from this commit)that we should do it as a separate patch.

One todo item is that we should add a behave test for table names with quotes and double-quotes too. I wanted to get feedback on this general approach before we spend more time doing that.

@entong, @addisonhuddy, what do you think?

### CODE REVIEW

File: gpMgmt/bin/analyzedb [hunk 1]
```diff
@@ -121,19 +125,22 @@ select distinct partitionschemaname, parentpartitiontablename from (%s) AS pps2
...
-SELECT count(*) FROM pg_attribute WHERE attrelid = '%s'::regclass AND attname IN (%s) AND attnum > 0 AND NOT attisdropped
+SELECT count(*) FROM pg_attribute WHERE attrelid = %s AND attname IN (%s) AND attnum > 0 AND NOT attisdropped
 """
 
 VALIDATE_TABLE_NAMES_SQL = """
-SELECT %s
+SELECT n.nspname, c.relname, orig_name
+FROM pg_class c
+INNER JOIN pg_namespace n ON c.relnamespace = n.oid
+INNER JOIN (VALUES %s) o(orig_name) ON c.oid = o.orig_name::regclass
```
Comment (entong) [inline gpMgmt/bin/analyzedb]: How's the performance of this query? I was using '%s'::regclass since it's very fast. We need to test that given a huge catalog and thousands of tables in the input config file, this does not become the bottleneck.

File: gpMgmt/bin/analyzedb [hunk 2]
```diff
@@ -281,12 +288,12 @@ class AnalyzeDb(Operation):
 
             logger.info("Checking for tables with stale stats...")
             # get all heap tables in the requested tables. all heap tables are regarded as dirty
-            heap_partitions = get_heap_tables_set(self.conn, input_tables_set) # set(['schema.table1', ...])
+            heap_partitions = get_heap_tables_set(self.conn, input_tables_set) # XXX set(['schema.table1', ...])
```
Comment (entong) [inline gpMgmt/bin/analyzedb]: I'm always curious what does the XXX mean in the code base.

File: gpMgmt/bin/analyzedb [hunk 3]
```diff
@@ -352,12 +361,18 @@ class AnalyzeDb(Operation):
             logger.info("Starting analyze with %d workers..." % num_workers)
             pool = AnalyzeWorkerPool(numWorkers=num_workers)
             for can in ordered_candidates:
+                can_schema = can[0]
+                can_table = can[1]
```
Comment (entong) [inline gpMgmt/bin/analyzedb]: can_schema, can_table = can[0], can[1]

File: gpMgmt/bin/analyzedb [hunk 4]
```diff
@@ -389,7 +404,8 @@ class AnalyzeDb(Operation):
                 logger.info("Total elapsed time: %d seconds. Analyzed %d out of %d table(s) or partition(s) successfully."
                             % (int(end_time-start_time), len(self.success_list), len(ordered_candidates)))
                 logger.info("Done.")
-        except:
+        except Exception, ex:
+            #logger.exception(ex)
```
Comment (entong) [inline gpMgmt/bin/analyzedb]: why this line is commented out?

File: gpMgmt/bin/analyzedb [hunk 5]
```diff
@@ -417,35 +433,47 @@ class AnalyzeDb(Operation):
         """
         logger.info("Getting and verifying input tables...")
         if self.single_table:
-            self.single_table = self.single_table.lower()
-            validate_include_tables(self.conn, [self.single_table], None)
+
+            # Check that the table name given on the command line is schema-qualified.
+            # XXX: Nowadays, the code should handle that just fine. Should we lift this limitation?
```
Comment (entong) [inline gpMgmt/bin/analyzedb]: This is to raise the awareness of the user that the name must be schema-qualified, since 9 out of 10 times the user will not do that the first time, especially if the schema is 'public'.

Comment (entong) [inline gpMgmt/bin/analyzedb]: I think we don't need to address every little corner case right now, which includes table names with a dot.

Comment (hlinnaka) [inline gpMgmt/bin/analyzedb]: On 21/12/15 22:07, Entong Shen wrote:
>> @@ -417,35 +433,47 @@ class AnalyzeDb(Operation):
>>           """
>>           logger.info("Getting and verifying input tables...")
>>           if self.single_table:
>> -            self.single_table = self.single_table.lower()
>> -            validate_include_tables(self.conn, [self.single_table], None)
>> +
>> +            # Check that the table name given on the command line is schema-qualified.
>> +            # XXX: Nowadays, the code should handle that just fine. Should we lift this limitation?
>
> This is to raise the awareness of the user that the name must be schema-qualified, since 9 out of 10 times the user will not do that the first time, especially if the schema is 'public'.

Well, the only reason the name must be schema-qualified is that we 
insist that it is. If we just removed that check from analyzedb, I think 
it would work fine (with this pull request).

- Heikki

File: gpMgmt/bin/analyzedb (hunk 4 above)
Comment (hlinnaka) [inline gpMgmt/bin/analyzedb]: On 21/12/15 22:04, Entong Shen wrote:
>> @@ -389,7 +404,8 @@ class AnalyzeDb(Operation):
>>                   logger.info("Total elapsed time: %d seconds. Analyzed %d out of %d table(s) or partition(s) successfully."
>>                               % (int(end_time-start_time), len(self.success_list), len(ordered_candidates)))
>>                   logger.info("Done.")
>> -        except:
>> +        except Exception, ex:
>> +            #logger.exception(ex)
>
> why this line is commented out?

Ah, leftover from debugging. Printing the traceback on exception was 
very helpful, but I didn't intend to include it in the final patch, as 
that's an unrelated change.

- Heikki

File: gpMgmt/bin/analyzedb (hunk 5 above)
Comment (hlinnaka) [inline gpMgmt/bin/analyzedb]: On 21/12/15 22:08, Entong Shen wrote:
>> @@ -417,35 +433,47 @@ class AnalyzeDb(Operation):
>>           """
>>           logger.info("Getting and verifying input tables...")
>>           if self.single_table:
>> -            self.single_table = self.single_table.lower()
>> -            validate_include_tables(self.conn, [self.single_table], None)
>> +
>> +            # Check that the table name given on the command line is schema-qualified.
>> +            # XXX: Nowadays, the code should handle that just fine. Should we lift this limitation?
>
> I think we don't need to address every little corner case right now, which includes table names with a dot.

Table names with dots actually work with this. E.g:

analyzedb -d postgres -t public.\"dot.table\"

Commas will however cause trouble, because comma is used as field 
separator in the state and last_op files, and there's no escaping there yet.

- Heikki

File: gpMgmt/bin/analyzedb (hunk 2 above)
Comment (hlinnaka) [inline gpMgmt/bin/analyzedb]: On 21/12/15 21:56, Entong Shen wrote:
>> @@ -281,12 +288,12 @@ class AnalyzeDb(Operation):
>>
>>               logger.info("Checking for tables with stale stats...")
>>               # get all heap tables in the requested tables. all heap tables are regarded as dirty
>> -            heap_partitions = get_heap_tables_set(self.conn, input_tables_set) # set(['schema.table1', ...])
>> +            heap_partitions = get_heap_tables_set(self.conn, input_tables_set) # XXX set(['schema.table1', ...])
>
> I'm always curious what does the XXX mean in the code base.

Ah. In this case, it means that the comment needs updating. 
get_heap_tables_set() doesn't return a set of strings anymore, but a set 
of (schema, table) tuples.

- Heikki

File: gpMgmt/bin/analyzedb (hunk 1 above)
Comment (hlinnaka) [inline gpMgmt/bin/analyzedb]: On 21/12/15 21:56, Entong Shen wrote:
>>   """
>>
>>   VALIDATE_TABLE_NAMES_SQL = """
>> -SELECT %s
>> +SELECT n.nspname, c.relname, orig_name
>> +FROM pg_class c
>> +INNER JOIN pg_namespace n ON c.relnamespace = n.oid
>> +INNER JOIN (VALUES %s) o(orig_name) ON c.oid = o.orig_name::regclass
>
> How's the performance of this query? I was using '%s'::regclass
> since it's very fast. We need to test that given a huge catalog and
> thousands of tables in the input config file, this does not become
> the bottleneck.

I doubt it's a problem, querying the catalogs should be cheap compared 
to actually performing the analyze, even if the query is badly 
optimized. But would be nice to test, of course.

- Heikki

**Comment by hlinnaka** [comment]
@entong, can you do the final cleanups/testing of this patch, or do you need further help with this?

File: gpMgmt/bin/analyzedb (hunk 1 above)
Comment (entong) [inline gpMgmt/bin/analyzedb]: I have a real schema loaded in a testing environment that has about 300K tables in the database. I'll manually test it over there and let you know.

File: gpMgmt/bin/analyzedb (hunk 5 above)
Comment (entong) [inline gpMgmt/bin/analyzedb]: Does the user need to escape double quotes in the command line? I hope at least for upper case letter in table names, they don't need to escape, i.e., analyzedb -d postgres -t public.Apple used to work without writing it like public."Apple" instead.

**Comment by entong** [comment]
@hlinnaka, thanks for the effort. I'll do the final cleanups and testing. It will take a couple of days since I'm actively working on sth else at the moment.

File: gpMgmt/bin/analyzedb (hunk 5 above)
Comment (hlinnaka) [inline gpMgmt/bin/analyzedb]: On 21/12/15 23:42, Entong Shen wrote:
>> @@ -417,35 +433,47 @@ class AnalyzeDb(Operation):
>>           """
>>           logger.info("Getting and verifying input tables...")
>>           if self.single_table:
>> -            self.single_table = self.single_table.lower()
>> -            validate_include_tables(self.conn, [self.single_table], None)
>> +
>> +            # Check that the table name given on the command line is schema-qualified.
>> +            # XXX: Nowadays, the code should handle that just fine. Should we lift this limitation?
>

> Does the user need to escape double quotes in the command line? I
> hope at least for upper case letter in table names, they don't need
> to escape, i.e., analyzedb -d postgres -t public.Apple used to work
> without writing it like public."Apple" instead.

Yes, they need to escape them. Same with upper-case letters. We could 
relax that for upper-case letters by putting in more code: we could 
check if the table name is of the form schema.CamelTable, where the 
table (or schema) contains upper-case letters, but doesn't otherwise 
need escaping, and turn it into schema."CamelTable".

I'm not sure if we should do change that, though, as long as this is 
documented; PostgreSQL's vacuumdb and reindexdb commands also behave the 
same, so there's some precedence for this. And it might surprise some 
users if it's case sensitive, while SQL in general isn't.

The schema name given with -s is currently case-sensitive though, and 
doesn't require escaping, so it's not very consistent as it is.

- Heikki

**Comment by hlinnaka** [comment]
On 21/12/15 23:44, Entong Shen wrote:
> @hlinnaka, thanks for the effort. I'll do the final cleanups and testing. It will take a couple of days since I'm actively working on sth else at the moment.

Great, thanks!

- Heikki

**Comment by addisonhuddy** [comment]
Awesome collaboration on this pull request. Great to see!

File: gpMgmt/bin/analyzedb [hunk 6]
```diff
@@ -473,23 +501,33 @@ class AnalyzeDb(Operation):
         2. check invalid characters in table names
         3. skip views and external tables
         """
+        qresult = run_sql(self.conn, GET_MID_LEVEL_PARTITIONS_SQL)
+        mid_level_partitions = []
+        for schema_tbl in qresult:
+            tuple = (schema_tbl[0], schema_tbl[1])
```
Comment (entong) [inline gpMgmt/bin/analyzedb]: I'll rename 'tuple' since it's a python keyword.

File: gpMgmt/bin/analyzedb [hunk 7]
```diff
@@ -473,23 +501,33 @@ class AnalyzeDb(Operation):
...
         if self.config_file is not None or self.single_table is not None:
             valid_tables = []
             if len(ret) > 0:
-                oid_str = ','.join(["'%s'::regclass" % x for x in ret])
+                oid_str = get_oid_str(ret)
                 qresult = run_sql(self.conn, GET_VALID_DATA_TABLES_SQL % oid_str)
-                for tbl in qresult:
-                    valid_tables.append('.'.join(tbl))
+                for schema_tbl in qresult:
+                    tuple = (schema_tbl[0], schema_tbl[1])
```
Comment (entong) [inline gpMgmt/bin/analyzedb]: same here

File: gpMgmt/bin/analyzedb [hunk 8]
```diff
@@ -508,20 +546,22 @@ class AnalyzeDb(Operation):
...
-        data = []
-        for row in rows:
-            line = "%s,%s,%d,%s,%s,%s" % (row[0], row[1], row[2], row[3], row[4], row[5])
-            data.append(line)
-        return data
+        oid_str = get_oid_str(input_tables_set)
+        qresult = run_sql(self.conn, GET_REQUESTED_LAST_OP_INFO_SQL % oid_str)
+        ret = []
+        for r in qresult:
+            tuple = (r[0], r[1], str(r[2]), r[3], r[4], r[5])
```
Comment (entong) [inline gpMgmt/bin/analyzedb]: same here

File: gpMgmt/bin/gppylib/operations/dump.py [hunk 9]
```diff
@@ -242,26 +252,6 @@ def get_last_state(table_type, master_datadir, backup_dir, dump_dir, dump_prefix
...
-
-def get_pgstatlastoperations_dict(last_operations):
-    last_operations_dict = {}
-    for operation in last_operations:
-        toks = operation.split(',')
-        if len(toks) != 6:
-            raise Exception('Wrong number of tokens in last_operation data for last backup: "%s"' % operation)
-        last_operations_dict[(toks[2], toks[3])] = operation
-    return last_operations_dict
-
```
Comment (entong) [inline gpMgmt/bin/gppylib/operations/dump.py]: I don't think these can be removed since there are used by other utilities like incremental backup.

File: gpMgmt/bin/gppylib/operations/dump.py [hunk 10]
```diff
@@ -207,10 +210,17 @@ def get_partition_state(master_port, dbname, catalog_schema, partition_info):
...
             validate_modcount(schemaname, partition_name, modcount)
-            partition_list.append('%s, %s, %s' %(schemaname, partition_name, modcount))
+            partition_list.append((schemaname, partition_name, modcount))
 
     return partition_list
 
+# A legacy version of get_partition_state_tuples() that returns a list of strings
+# instead of tuples. Should not be used in new code, because the string
+# representation doesn't handle schema or table names with commas.
+def get_partition_state(master_port, dbname, catalog_schema, partition_info):
```
Comment (entong) [inline gpMgmt/bin/gppylib/operations/dump.py]: I'll double check if this function returns the same as before, since it is used elsewhere besides analyzedb.
